File: C/2/part-1-problem/part-1-problem.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h> // Required for strtok

#define MAX_LINE_LENGTH 1024

// Function to check if a row is "safe"
bool isSafeRow(int *row, int size) {
    // Check consecutive differences
    for (int i = 0; i < size - 1; i++) {
        int diff = abs(row[i + 1] - row[i]);
        if (diff < 1 || diff > 3) {
            return false; // Unsafe condition
        }
    }

    // Check monotonicity
    bool increasing = true, decreasing = true;
    for (int i = 0; i < size - 1; i++) {
        if (row[i] < row[i + 1]) {
            decreasing = false;
        } else if (row[i] > row[i + 1]) {
            increasing = false;
        }
    }

    if (!increasing && !decreasing) {
        return false; // Not monotonic, hence unsafe
    }

    return true; // Safe if all conditions pass
}

// Function to process the file and determine "safe" rows
void processFile(const char *filename) {
    FILE *file = fopen(filename, "r");
    if (file == NULL) {
        perror("Error opening file");
        return;
    }

    char line[MAX_LINE_LENGTH];
    int safeCount = 0, unsafeCount = 0;

    while (fgets(line, MAX_LINE_LENGTH, file)) {
        int row[MAX_LINE_LENGTH / 2]; // Temporary array for storing numbers
        int size = 0;

        // Parse numbers from the line
        char *token = strtok(line, " \t\n");
        while (token != NULL) {
            row[size++] = atoi(token);
            token = strtok(NULL, " \t\n");
        }

        // Check if the row is safe
        if (isSafeRow(row, size)) {
            safeCount++;
        } else {
            unsafeCount++;
        }
    }

    fclose(file);

    // Output the results
    printf("Total safe rows: %d\n", safeCount);
    printf("Total unsafe rows: %d\n", unsafeCount);
}

int main() {
    char filename[100];
    printf("Enter the name of the input file: ");
    scanf("%s", filename);

    processFile(filename);

    return 0;
}

File: 2024/Go/2/part-1-problem/part-1-problem.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

// Function to check if a row is "safe"
func isSafeRow(row []int) bool {
	// Check consecutive differences
	for i := 0; i < len(row)-1; i++ {
		diff := int(math.Abs(float64(row[i+1] - row[i])))
		if diff < 1 || diff > 3 {
			return false // Unsafe condition
		}
	}

	// Check monotonicity
	increasing, decreasing := true, true
	for i := 0; i < len(row)-1; i++ {
		if row[i] < row[i+1] {
			decreasing = false
		} else if row[i] > row[i+1] {
			increasing = false
		}
	}

	if !increasing && !decreasing {
		return false // Not monotonic, hence unsafe
	}

	return true // Safe if all conditions pass
}

// Function to process the file and determine "safe" rows
func processFile(filename string) {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	safeCount, unsafeCount := 0, 0

	for scanner.Scan() {
		// Read the line and split it into numbers
		line := scanner.Text()
		tokens := strings.Fields(line)
		row := make([]int, len(tokens))

		// Convert tokens to integers
		for i, token := range tokens {
			num, err := strconv.Atoi(token)
			if err != nil {
				fmt.Println("Error parsing number:", err)
				return
			}
			row[i] = num
		}

		// Check if the row is safe
		if isSafeRow(row) {
			safeCount++
		} else {
			unsafeCount++
		}
	}

	// Output results
	fmt.Printf("Total safe rows: %d\n", safeCount)
	fmt.Printf("Total unsafe rows: %d\n", unsafeCount)

	// Check for scanning errors
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
	}
}

func main() {
	var filename string
	fmt.Print("Enter the name of the input file: ")
	fmt.Scan(&filename)

	processFile(filename)
}
